### PR TITLE
Add some more command arguments

### PR DESCRIPTION
**Pull request by pie-flavor**
I figured the things in MoreArguments would be better used in the standard library.

[Test plugin as requested by](https://github.com/pie-flavor/SpongeTestPlugin/releases/tag/more-arguments%2F1.0.1) @kashike

### CODE REVIEW

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (me4502) [inline src/main/java/org/spongepowered/api/command/args/GenericArguments.java]: Configurate supports numerous configuration types, even YAML. Also I don't see the appeal of creating a ConfigurationNode. It'd make more sense to just have a generic JSON deserializing argument.

Comment (pie-flavor) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java]: The appeal is it's already there, with wide use already, and there's also the bit about object mapping. And if not a node, what would it return?

And the idea here is that "Mojangson" is unintentionally a subset of HOCON, and other commands already use strict JSON.

I wouldn't know how to actually make a JSON/Mojangson deserializing argument; I just know how to use the Configurate library. If it shouldn't use Configurate, I'd just remove it.

Comment (me4502) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java]: I'd probably ask @zml2008 what he thinks should work.

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/command/args/GenericArguments.java]: Why do you place the if here? Do you want to verify that it matches the URI format even if it is an URL?
I guess you could move the if before the uri part or wrap the toURI part in the if. Thats just my thought though.

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/command/args/GenericArguments.java]: How about specifing a player name here to obtain the ip:
`/banip badGuy`

Comment (ryantheleach) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java]: UUID would be nice too.

Comment (pie-flavor) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java]: That sounds fun.

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (pie-flavor) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java]: Yeah, the goal here is to verify the structure either way but return what they're asking for. I wasn't about to pull in all of Apache Validators just for the UrlValidator.

Comment (windy1) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java]: From the JavaDocs:
> Note, any URL instance that complies with RFC 2396 can be converted to a URI. However, some URLs that are not strictly in compliance can not be converted to a URI.

I think it's best to maybe validate that the URL is compliant by calling `toURI` but then ultimately returning the `URL`. Returning one or the other is too ambiguous in my eyes here.

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (windy1) [inline src/main/java/org/spongepowered/api/command/args/GenericArguments.java:1824]: I like this idea and I think it could be applied elsewhere as well (e.g. `entityOrTarget`).

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (windy1) [inline src/main/java/org/spongepowered/api/command/args/GenericArguments.java:1849]: Correct me if I'm wrong here but wouldn't this get the index of the argument you are parsing, therefore making your call to `args.setState` reset the index to before the argument you've already parsed? Also is dealing with the state even necessary for single argument `CommandElement`s? I was under the impression that it is used for multi-arg elements such as `remainingJoinedStrings`.

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (windy1) [inline src/main/java/org/spongepowered/api/command/args/GenericArguments.java:1879]: Hmm, I don't know about this one, can you describe a use case where a regular `double` is insufficient for user input? I understand the implications of float point numbers in calculation, but for user input?

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (windy1) [inline src/main/java/org/spongepowered/api/command/args/GenericArguments.java:1907]: This one is a bit more understandable I suppose, still a little questionable, IMO.

**Review by windy1 — CHANGES_REQUESTED** [review]
Looks good overall, definitely justified in most cases IMO.

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (pie-flavor) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java:1824]: Hmm? We already have `entityOrSource`.

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (pie-flavor) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java:1879]: Actually, it's `bigInteger` that is somewhat unnecessary - the point of `bigDecimal` is that the Economy API is entirely structured around them, and I find myself constantly using casts and checks (not to mention the fact that an arbitrary economy plugin may use enormous numbers).

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (pie-flavor) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java:1849]: That is in fact the point; otherwise this argument would be swallowed without parsing if the "or source" part is executed.

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (windy1) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java:1879]: Right, but `BigDecimal` has a `double` value constructor and the point of `BigDecimal` is more about it's increased precision which I don't think is really applicable to humans in general and is more meant for lossless floating point calculations.

Comment (pie-flavor) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java:1879]: Well, an arbitrary economy API might use incredibly large precise values. Humans are humans, but command blocks exist as well. And the point here is abstraction; why shouldn't you be able to request the one thing that half of your plugin is structured around?

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (pie-flavor) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java]: I suppose.

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (windy1) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java:1824]: Oh did not know that, I still think an `entityOrTarget` (by directional vector) would be cool.

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (sibomots) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java:1879]: Generics, man. Generics.   fooTypeX, fooTypeY, fooTypeZ is not generic.

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (sibomots) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java]: the name of this function tells me nothing specific.  The name says "configNode". Does it config _a_ node? which node, a parameter.. nope the object's node, well it's static method of a "GenericArguments" class.

**Comment by sibomots** [comment]
The name of this PR is unclear.  Is there a style guide floating around for how to name bugs/PR ?

I'll name my next PR "Stuff that works better."

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (phroa) [inline src/main/java/org/spongepowered/api/command/args/GenericArguments.java]: According to the javadoc for `ipOrSource` you meant `RemoteSource` here and in the cast below.

**Review by phroa — CHANGES_REQUESTED** [review]
...I'm not sure "start a review" was what I should have clicked.  I guess I'll request changes?

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (pie-flavor) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java]: That is a good point - it's not exactly clear that it's a 'config object'. Any input on new names?

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (pie-flavor) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java:1824]: That's actually a very good idea, I'll look into that. That's one that would get used plenty (and by Reveries, no less :P) 

Edit: Sorta used the exact code

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (pie-flavor) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java:1879]: If only `GenericArguments.of(BigInteger.class)` was feasible.

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (ryantheleach) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java:1824]: There is a #target hard coded entity name for selecting entity targets I think. Running off memory due to xmas

Comment (pie-flavor) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java:1824]: Not for entities, just for vectors/locations.

**Comment by pie-flavor** [comment]
@windy1 Are you still requesting changes? Because it seems like it was resolved but the change request is still open.

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/command/args/GenericArguments.java]: Shouldn't this method be above `entityOrSource()`? In order to have similar methods are next to each other.
In addition to that the `entityOrSource(Text key, Class<? extends Entity> type)` variant is missing.

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/command/args/GenericArguments.java:1969]: Maybe there should be a helper method, that allows some uuid suggestions (online players/entities etc).

````
public static CommandElement uuid(Text key, Collection<UUID> suggetedUUIDs);
````

Or is there a CommandElement extension that allows adding custom suggestions? <- Maybe we should add one.

**Review by ST-DDT — CHANGES_REQUESTED** [review]

**Comment by ST-DDT** [comment]
@pie-flavor The system will not recognize it as resolved until @windy1 reviews your PR again and presses the accept button.

**Comment by pie-flavor** [comment]
@ST-DDT I know, that's why I poked him.

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (pie-flavor) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java:1969]: The point here is that it's a plain UUID, not related to anything in particular. `entity()` and `player()` methods already exist should they want it to fit a pattern.

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (pie-flavor) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java]: I appended to the bottom; if this is an issue I'll move it.

Intentional; if `orSource` is used, the target type is a player.

Comment (ST-DDT) [reply src/main/java/org/spongepowered/api/command/args/GenericArguments.java]: > I appended to the bottom; if this is an issue I'll move it.

I just mentioned it as my own preference. Whether you should change it or not is up to you/others.

> Intentional; if `orSource` is used, the target type is a player.

Well I thought the class limitation would be `Entity <= X <= Player`.

This would allow things such as:
`/heal [Living/Player]`

But I guess there are other ways as well.

**Comment by pie-flavor** [comment]
@windy1 Still waiting on that clarification :P

File: src/main/java/org/spongepowered/api/command/args/GenericArguments.java
Comment (parlough) [inline src/main/java/org/spongepowered/api/command/args/GenericArguments.java]: Probably should document this and the two methods below(`entity(Text key, Class<? extends Entity> type)`, `entityOrTarget(Text key)`, and `entityOrTarget(Text key, Class<? extends Entity> type)` , explaining what they accomplish.

As a whole this PR seems good and I would love to see it merged.

**Comment by kashike** [comment]
@pie-flavor status?

**Comment by pie-flavor** [comment]
@kashike status is if you can't find anything wrong then it's ready, but still waiting for windy to approve changes

**Comment by kashike** [comment]
@pie-flavor Could you provide a test plugin in the description?

**Comment by pie-flavor** [comment]
@windy1 any updates?

**Comment by pie-flavor** [comment]
Um.
Is that commit going to create problems? I ask because I still don't fully understand the merge conflict thing, and I used GitHub's tool this time.

**Comment by Deamon5550** [comment]
Its usually better to rebase than merge when updating feature branches

**Comment by pie-flavor** [comment]
um

approximately 0.01% of that was intentional.

**Comment by pie-flavor** [comment]
Right, I give up. GG git.

**Pull request closed without merging**
